File: Lab-4-lambda/node_10.18.1/get_ratings_code.js
var 
    AWS_SDK = require("aws-sdk"),
    FS = require("fs"),
    S3_API = new AWS_SDK.S3({
        region: "us-east-1",
        apiVersion: "2006-03-01"
    }),
    SOURCE_S3_BUCKET_STR = "aws-tc-largeobjects",
    SOURCE_KEY_STR = "DEV-AWS-MO-Building_2.0/my_json_lines.jsonl";

exports.handler = function(event, context, callback){
    var
        G_PRODUCT_ID_STR = event.product_id_str,
        SEARCH_EXPRESSION_STR = "SELECT s.star_rating FROM S3Object s WHERE s.product_id = '" + G_PRODUCT_ID_STR + "'",
        g_return_me = {
            "statusCode": 200
        },
        g_search_results_arr = [],
        g_average_float_arr = [];

    (function init(){
        doSearch(function(err, chunk_str){
            if(err){
                return callback(err, null);
            }
            processStitchedChunks(chunk_str);
            g_return_me = constructBody();
            callback(null, g_return_me);
        });
    })();

    function constructBody(){
        var 
            body = {};
        body.product_id_str = event.product_id_str;
        body.average_star_review_float = _returnAv();
        return body;
    }

    function _returnAv(){
        var 
            sum_float = g_average_float_arr.reduce((a, b) => a + b, 0),
            avg_float = (sum_float / g_average_float_arr.length) || 0;
       return avg_float;
    }
    function processStitchedChunks(stiched_chunk_str){
        var 
            line_arr = stiched_chunk_str.split("\n").filter(Boolean);
        for(var i_int = 0; i_int < line_arr.length; i_int += 1){
            var
                o = JSON.parse(line_arr[i_int]);
             g_search_results_arr.push(o);
             g_average_float_arr.push(Number(o.star_rating));
        }
        console.log(g_search_results_arr.length.toString() + " records found");
    }

    function doSearch(cb){
        var 
            params = {
                Bucket: SOURCE_S3_BUCKET_STR,
                Key: SOURCE_KEY_STR,
                Expression: SEARCH_EXPRESSION_STR,
                ExpressionType: "SQL",
                InputSerialization: {
                    JSON: {
                        Type: "LINES"
                    },
                },
                OutputSerialization: {
                    JSON: {

                    }
                }
            };
        // await/async support in beta for selectObjectContext and the api is not working correcty.
        // using a stadard cb approach
        S3_API.selectObjectContent(params, function(err, data){
            var chunk_str = "";
            if(err){
                console.log("err", err);
                return;
            }
            var event = data.Payload;
            event.on("data", function(event) {
                if(event.Records){
                    chunk_str += event.Records.Payload.toString();
                }
            });
            event.on("error", function(err){
                cb(err, null);
            });
            event.on("end", function(){ 
                cb(null, chunk_str);
            });
        });
    }
};

File: Lab-4-lambda/python_3.6.8/get_ratings_code.py
import boto3
import json
import statistics
import re

def lambda_handler(event, context):
    PRODUCT = event['product_id_str']
    S3_BUCKET = 'aws-tc-largeobjects'
    S3_FILE = 'DEV-AWS-MO-Building_2.0/my_json_lines.jsonl'

    s3 = boto3.client('s3')

    r = s3.select_object_content(
            Bucket=S3_BUCKET,
            Key=S3_FILE,
            ExpressionType='SQL',
            Expression="select s.star_rating from s3object[*] s where s.product_id = '" + PRODUCT + "'",
            InputSerialization={'JSON': {"Type": "Lines"}},
            OutputSerialization={'JSON': {}}
    )

    for event in r['Payload']:
        if 'Records' in event:
            records = event['Records']['Payload'].decode('utf-8')
            n_records = (re.findall(r'\d+', records))
            results = list(map(float, n_records))
            f_results = statistics.mean(results)
            
    return {
        'product_id_str': PRODUCT,
        'average_star_review_float': f_results
    }